File: day10/src/main/java/de/havox_design/aoc2016/day10/Day10.java
package de.havox_design.aoc2016.day10;

import de.havox_design.aoc2016.utils.DataReader;

import java.util.List;

public class Day10 {
    private final List<String> input;

    public Day10(String fileName) {
        input = readData(fileName);
    }

    public static long solvePart1(String fileName) {
        Day10 instance = new Day10(fileName);
        return instance.solvePart1();
    }

    public static long solvePart2(String fileName) {
        Day10 instance = new Day10(fileName);
        return instance.solvePart2();
    }

    public long solvePart1() {
        return 0L;
    }

    public long solvePart2() {return 0L;
    }

    private List<String> readData(String fileName) {
        return DataReader.readData(fileName, MainClass.class);
    }
}

File: day10/src/main/scala/de/havox_design/aoc2017/day10/Day10.scala
package de.havox_design.aoc2017.day10

object Day10 {
  def solvePart1(filename: String, elements: Int = 256): Int =
    calculateProductOfFirstTwoElements(readData(filename).next(), elements)

  def solvePart2(filename: String): String =
    ""

  def main(args: Array[String]): Unit = {
    def dayFileName = "day07.txt"

    println("Solution for part1: " + solvePart1(dayFileName))
    println("Solution for part2: " + solvePart2(dayFileName))
  }

  private def process(initialKnotState: KnotState, lengths: Seq[Int]): KnotState =
    lengths
      .foldLeft(initialKnotState)(_.reversed(_))

  private def calculateProductOfFirstTwoElements(input: String, elements: Int = 256): Int = {
    val lengths = input
      .trim
      .split(",")
      .toSeq
      .map(
        _
          .trim
          .toInt
      )

    process(KnotState(Vector.range(0, elements)), lengths)
      .checkProduct
  }

  private def readData(filename: String): Iterator[String] =
    scala
      .io
      .Source
      .fromResource(filename)
      .getLines()
}
